src/com/trilead/ssh2/crypto/digest/HMAC.java

```diff
@@ -1,96 +1,166 @@
 
 package com.trilead.ssh2.crypto.digest;
 
+import javax.crypto.Mac;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
 /**
- * HMAC.
- * 
+ * MAC.
+ *
  * @author Christian Plattner, plattner@trilead.com
- * @version $Id: HMAC.java,v 1.1 2007/10/15 12:49:57 cplattne Exp $
+ * @version $Id: MAC.java,v 1.1 2007/10/15 12:49:57 cplattne Exp $
  */
-@Deprecated
-public final class HMAC implements Digest
+public final class HMAC implements MAC
 {
-	Digest md;
-	byte[] k_xor_ipad;
-	byte[] k_xor_opad;
-
-	byte[] tmp;
-
-	int size;
-
-	public HMAC(Digest md, byte[] key, int size)
-	{
-		this.md = md;
-		this.size = size;
-
-		tmp = new byte[md.getDigestLength()];
-
-		final int BLOCKSIZE = 64;
-
-		k_xor_ipad = new byte[BLOCKSIZE];
-		k_xor_opad = new byte[BLOCKSIZE];
-
-		if (key.length > BLOCKSIZE)
-		{
-			md.reset();
-			md.update(key);
-			md.digest(tmp);
-			key = tmp;
-		}
-
-		System.arraycopy(key, 0, k_xor_ipad, 0, key.length);
-		System.arraycopy(key, 0, k_xor_opad, 0, key.length);
-
-		for (int i = 0; i < BLOCKSIZE; i++)
-		{
-			k_xor_ipad[i] ^= 0x36;
-			k_xor_opad[i] ^= 0x5C;
-		}
-		md.update(k_xor_ipad);
-	}
-
-	public final int getDigestLength()
-	{
-		return size;
-	}
-
-	public final void update(byte b)
-	{
-		md.update(b);
-	}
-
-	public final void update(byte[] b)
-	{
-		md.update(b);
-	}
-
-	public final void update(byte[] b, int off, int len)
-	{
-		md.update(b, off, len);
-	}
-
-	public final void reset()
-	{
-		md.reset();
-		md.update(k_xor_ipad);
-	}
-
-	public final void digest(byte[] out)
-	{
-		digest(out, 0);
-	}
-
-	public final void digest(byte[] out, int off)
-	{
-		md.digest(tmp);
-
-		md.update(k_xor_opad);
-		md.update(tmp);
-
-		md.digest(tmp);
-
-		System.arraycopy(tmp, 0, out, off, size);
-
-		md.update(k_xor_ipad);
-	}
+    private static final String ETM_SUFFIX = "-etm@openssh.com";
+
+    /**
+     * From http://tools.ietf.org/html/rfc4253
+     */
+    static final String HMAC_MD5 = "hmac-md5";
```

Comment:
Anything with MD5 in it is also old enough to remove. SHA1 is on the way out as well.

Reply:
@jvz  could it be done in a separate review?

Reply:
Yes, if this plugin continues development maintenance, then removal or disabling of old, weak algorithms can be done in a standalone PR. I originally left these comments when the PR was practically rewriting the library which seemed like an opportune time to clean up legacy cruft. :)